frame/elections-phragmen/src/benchmarking.rs

```diff
@@ -0,0 +1,283 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Elections-Phragmen pallet benchmarking.
+
+#![cfg(feature = "runtime-benchmarks")]
+
+use super::*;
+
+use codec::{Encode, Decode};
+use frame_system::RawOrigin;
+use sp_io::hashing::blake2_256;
+use frame_benchmarking::benchmarks;
+use sp_runtime::traits::Bounded;
+use frame_support::{assert_ok, traits::OnInitialize};
+
+use crate::Module as Elections;
+
+const SEED: u32 = 0;
+const BALANCE_FACTOR: u32 = 250;
+
+type Lookup<T> = <<T as frame_system::Trait>::Lookup as StaticLookup>::Source;
+
+/// grab a new account
+fn account<T: Trait>(name: &'static str, index: u32) -> T::AccountId {
+	let entropy = (name, index).using_encoded(blake2_256);
+	T::AccountId::decode(&mut &entropy[..]).unwrap_or_default()
+}
+
+/// grab new account with infinite balance.
+fn endowed_account<T: Trait>(name: &'static str, index: u32) -> T::AccountId {
+	let account = account::<T>(name, index);
+	let _ = T::Currency::make_free_balance_be(&account, BalanceOf::<T>::max_value());
+	account
+}
+
+/// Account ot lookup type of system trait.
+fn as_lookup<T: Trait>(account: T::AccountId) -> Lookup<T> {
+	T::Lookup::unlookup(account)
+}
+
+/// Get a reasonable amount of stake based on the execution trait's configuration
+fn default_stake<T: Trait>(factor: u32) -> BalanceOf<T> {
+	let factor = BalanceOf::<T>::from(factor);
+	T::Currency::minimum_balance() * factor
+}
+
+/// Add `c` new candidates.
+fn submit_candidates<T: Trait>(c: u32) -> Result<Vec<T::AccountId>, &'static str> {
+	(0..c).map(|i| {
+		let account = endowed_account::<T>("candidate", i);
+		<Elections<T>>::submit_candidacy(RawOrigin::Signed(account.clone()).into())
+			.map_err(|_| "failed to submit candidacy")?;
+		Ok(account)
+	}).collect::<Result<_, _>>()
+}
+
+/// Submit one voter.
+fn submit_voter<T: Trait>(caller: T::AccountId, votes: Vec<T::AccountId>, stake: BalanceOf<T>)
+	-> Result<(), &'static str>
+{
+	<Elections<T>>::vote(RawOrigin::Signed(caller).into(), votes, stake)
+		.map_err(|e| "failed to submit vote")
+}
+
+benchmarks! {
+	_ {}
+
+	// -- Signed ones
+	vote {
+		// range of candidates to vote for. Direct argument of the dispatchable.
+		let x in 1 .. (MAXIMUM_VOTE as u32);
```

Comment:
`vote` extrinsic depends on the length of `candidates` and `members` right?

```
let candidates_count = <Candidates<T>>::decode_len().unwrap_or(0) as usize;
let members_count = <Members<T>>::decode_len().unwrap_or(0) as usize;
```

Also depends on the number of locks on a user, because you set a new lock.

Reply:
depends in which sense? you can vote a number `x` where `(candidates + members + runners)< x < MAXIMUM_VOTE `. And assuming the worse case, the bound is always `MAXIMUM_VOTE`, i.e. we have more candidates and members than `MAXIMUM_VOTE`. The user can then vote for any number `1 < x < MAXIMUM_VOTE`.

and I am not aware of how locking is related here. Indeed we set a new lock here. So it matters if my user had previously 1 locks or 100 locks?

Reply:
Okay I got the deal with locks. 

but members and candidates, I disagree. It does not matter if we have members, candidates, or runners-up. We can vote for at most `MAXIMUM_VOTE`, regardless of how many members etc. we have. The onle case is if sum of these 3 are less than `MAXIMUM_VOTE `, which I safely ignore.